File: npro/npro/report/leads_by_status_owner/leads_by_status_owner.py
# Copyright (c) 2013, GreyCube Technologies and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
import pandas


def execute(filters=None):
    return get_data(filters)


def get_data(filters):

    columns = [
        dict(label="Sales Rep", fieldname="lead_owner", fieldtype="Data", width=165)
    ]

    data = frappe.db.sql(
        """
        select
            u.full_name lead_owner, l.status, count(l.status) total_count
        from
            tabLead l
            inner join tabUser u on u.name = l.lead_owner
        {where_conditions}
        group by 
            lead_owner, status""".format(
            where_conditions=get_conditions(filters)
        ),
        filters,
        as_dict=True,
        # debug=True,
    )
    if not data:
        return columns, []

    df = pandas.DataFrame.from_records(data)
    df1 = pandas.pivot_table(
        df,
        index=["lead_owner"],
        values=["total_count"],
        columns=["status"],
        aggfunc=sum,
        fill_value=0,
        margins=True,
    )
    df1.drop(index="All", axis=0, inplace=True)
    df1.columns = [d for d in df1.columns.to_series().str[1]]
    df2 = df1.reset_index()
 
    #  sorting grid columns
    sort_order = (
        frappe.db.get_single_value("NPro Settings", "lead_status_sort_order") or ""
    )
    sort_order = sort_order.split(",")
    columns = columns + sorted([
        dict(label=frappe.unscrub(col), fieldname=col, fieldtype="Int", width=95)
        for col in df1.columns],
        key=lambda x: sort_order.index(x) if x in sort_order
        else 100,
    )

    return columns, df2.to_dict("r")


def get_conditions(filters):
    conditions = []
    if filters.get("from_date"):
        conditions += ["date(l.creation) >= %(from_date)s"]
    if filters.get("to_date"):
        conditions += ["date(l.creation) <= %(to_date)s"]

    return conditions and " where " + " and ".join(conditions) or ""

File: npro/npro/report/leads_by_status_owner/leads_by_status_owner.js
// Copyright (c) 2016, GreyCube Technologies and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Leads By Status Owner"] = {
  filters: [
    {
      fieldname: "from_date",
      label: __("From Date (Lead Creation Date)"),
      fieldtype: "Date",
      default: moment().startOf("month").subtract(3, "months"),
    },
    {
      fieldname: "to_date",
      label: __("To Date (Lead Creation Date)"),
      fieldtype: "Date",
      default: frappe.datetime.get_today(),
    },
    {
      fieldname: "timespan",
      label: __("Lead Creation Date in "),
      fieldtype: "Select",
      options: npro.utils.TIMESPAN_OPTIONS,
      on_change: function (query_report) {
        let date_range = npro.utils.get_date_range(
          query_report.get_values().timespan
        );
        frappe.query_report.set_filter_value({
          from_date: date_range[0],
          to_date: date_range[1],
        });
      },
      default: "This Month",
    },
  ],

  _after_datatable_render: function (datatable) {
    const chart_columns = {
      New: "#fff168",
      Working: "#aee4ff",
      Nurturing: "#9deca2",
      Converted: "#49937E",
      "Do Not Contact": "#ff4d4d",
    };
    npro.utils.create_chart(
      "lead_owner",
      chart_columns,
      datatable,
      frappe.query_report
    );
  },
};
